File: sdat.R
library(httr)
#library(RSocrata)
library(tidyverse)
library(lubridate)
#slibrary(stringr)
library(caret)

api_endpoint = "https://opendata.maryland.gov/resource/ed4q-f8tm.json?"
api_csv = "https://opendata.maryland.gov/resource/ed4q-f8tm.csv?"

url_md = "https://opendata.maryland.gov/"
app_id = "fuixaI8OmedLW38WcI00DnoCn"

select_statement = "SELECT 
    mdp_street_address_mdp_field_address as address,
    mdp_street_address_zip_code_mdp_field_zipcode as zipcode,
    mdp_street_address_city_mdp_field_city as city,
    subdivision_code_mdp_field_subdivsn_sdat_field_37 as neighborhood,
    record_key_owner_occupancy_code_mdp_field_ooi_sdat_field_6 as owner_type,
    sales_segment_1_consideration_mdp_field_considr1_sdat_field_90 as price,
    sales_segment_2_consideration_sdat_field_110 as price2,
    sales_segment_1_transfer_date_yyyy_mm_dd_mdp_field_tradate_sdat_field_89 as date,
    sales_segment_2_transfer_date_yyyy_mm_dd_sdat_field_109 as date2,

    sales_segment_1_grantor_name_mdp_field_grntnam1_sdat_field_80 as seller,
    c_a_m_a_system_data_year_built_yyyy_mdp_field_yearblt_sdat_field_235 as year_built,
    c_a_m_a_system_data_dwelling_grade_code_and_description_mdp_field_strugrad_strudesc_sdat_field_230 as grade,
    current_assessment_year_total_assessment_sdat_field_172 as tax_assessment,
    sales_segment_3_how_conveyed_ind_sdat_field_127 as sales_type_orig,
    c_a_m_a_system_data_structure_area_sq_ft_mdp_field_sqftstrc_sdat_field_241 as living_area,
    c_a_m_a_system_data_land_area_mdp_field_landarea_sdat_field_242 as land_area,
    c_a_m_a_system_data_land_unit_of_measure_mdp_field_luom_sdat_field_243 as land_area_measure,
    land_use_code_mdp_field_lu_desclu_sdat_field_50 as land_use,
    additional_c_a_m_a_data_dwelling_construction_code_mdp_field_strucnst_sdat_field_263 as constr_code,
    additional_c_a_m_a_data_building_style_code_and_description_mdp_field_strustyl_descstyl_sdat_field_264 as bld_code_detail,
    additional_c_a_m_a_data_dwelling_type_mdp_field_strubldg_sdat_field_265 as bld_code,
    latitude_longitude as location,
    mdp_latitude_mdp_field_digycord_converted_to_wgs84 as lat,
    mdp_longitude_mdp_field_digxcord_converted_to_wgs84 as lon
"

# additional_c_a_m_a_data_building_style_code_and_description_mdp_field_strustyl_descstyl_sdat_field_1 as style_code,

select_statement = gsub("[\r\n]", "",select_statement)
select_statement = gsub("  ", "",select_statement)


where_comps = function(lat, lon, miles = 1, year = 2019){ 
  miles = miles/0.000621371 #convert to meters
  w = sprintf(" WHERE 
        
        sales_segment_1_transfer_date_yyyy_mm_dd_mdp_field_tradate_sdat_field_89 >= %s  AND 
        sales_segment_1_consideration_mdp_field_considr1_sdat_field_90 > 10000 AND 
        within_circle(latitude_longitude, %f, %f, %f)", paste0("'", year, "%'"),lat, lon, miles)
  w = gsub("[\r\n]", "",w)
  w = gsub("  ", "",w)
  return(w)
  # (sales_segment_1_consideration_mdp_field_considr1_sdat_field_90 - sales_segment_2_consideration_sdat_field_110) > 50000 AND 
  # sales_segment_2_consideration_sdat_field_110 > 0 AND 

}

clean_addresses <- function(addresses){
  props_str = toupper(addresses)
  props_str = str_replace_all(props_str,"STREET", "ST")
  props_str = str_replace_all(props_str,"AVENUE", "AVE")
  props_str = str_replace_all(props_str,"DRIVE", "DR")
  props_str = str_replace_all(props_str,"LN", "LANE")
  props_str = str_replace_all(props_str,"TERRACE", "TER")
  props_str = str_replace_all(props_str,"COURT", "CT")
  props_str = str_replace_all(props_str," RD", " ROAD")
  props_str = str_replace_all(props_str,"TRAIL", "TRL")
  props_str = str_replace_all(props_str,"PLACE", "PL")
  props_str = str_replace_all(props_str,"BOULEVARD", "BLVD")
  
  return(props_str)
}

where_meta = function(props){
  props = paste0("'", sapply(props, trimws), "'")
  props_str = paste0(props, collapse = ",")
  props_str = clean_addresses(props_str)
  
  
  
  
  w = sprintf(" WHERE mdp_street_address_mdp_field_address in (%s) ", props_str)
  w = gsub("[\r\n]", "",w)
  w = gsub("  ", "",w)
  return(w)
}

sdat_query = function(api=api_csv, select=select_statement, where,limit = 50000,return_page=0){
  query = paste0("$query= ", select, where, " LIMIT ", limit)
  full_url = paste0(api, query)
  full_url = URLencode(full_url)
  page = GET(full_url)
  if(return_page>0){return(page)}
  if(status_code(page)==200){
    frame = suppressMessages(content(page))
    
    #clean data
    try({
      frame$neighborhood <- frame$neighborhood %>% as.factor()
      frame <- filter(frame, living_area >500)
      frame$price_delta <- (frame$price - frame$price2)
      frame$date_delta <- (ymd(frame$date) - ymd(frame$date2))/ 365
      frame$renovated <- frame$price_delta > 50000 & frame$date_delta <=1 & frame$price2 >10000
      frame$basement <- grepl("with basement", tolower(frame$bld_code_detail))
      frame$stories <- sub(".*STRY *(.*?) *Story.*", "\\1", frame$bld_code_detail)
      frame$acre <- ""
      frame$acre[frame$land_area_measure =="S"] <- frame$land_area[frame$land_area_measure =="S"] / 43560
      frame$acre[frame$land_area_measure =="A"] <- frame$land_area[frame$land_area_measure =="A"]
      
    })
    return(frame)
  }else{
    return(NULL)
  }
  
}

sdat_q=function(api=api_csv, q){
  query= paste0("$q=", q)
  full_url = paste0(api, query)
  full_url = URLencode(full_url)
  page = GET(full_url)
  if(status_code(page)==200){
    frame = content(page)
    return(frame)
  }else{
    return(NULL)
  }
  
}
sdat_get_meta <- function(df, col = "address"){
  results <- NULL
  for(i in 1:nrow(df)){
    tmp_result <- NULL
    tmp_result <- sdat_query(where=where_meta(df[,col][i]))
    if(!is.null(tmp_result)){
      #tmp_result$search_address <- df[,col][i] 
      if(is.null(results)){
        results <- tmp_result
      }else{
        results <- rbind(results, tmp_result)
    }
  }
  }
  df$search_address <- clean_addresses(df[,col])
  df <- left_join(df, results, by=c("search_address" = "address"))
  return(df)
}

sdat_get_many_properties = function(addresses){
  n = length(addresses)
  result = NULL
  beg = 0
  end = 0
  while(end<n){
    beg= beg+1
    end = min(beg +9 , n)
    print(paste("beginning", beg))
    print(paste("END", end))
    tmp_result = sdat_query(where=where_meta(addresses[beg:end]))
    if(is.null(result)){
      result = tmp_result
    }else{
      result = rbind(result, tmp_result)
    }
    tmp_result = NULL   
    beg = end
  }
  return(result)
}

model.lm <- function(df, reno = T, dist = 1){
  if(reno){df <- filter(df, renovated==T)}
  
  lm(price~living_area + stories + basement , data = df)
}

model.rf <- function(df){
  #control <- trainControl(method='boot', 
  #                        number=10, 
  #                       repeats=3)
  train(price~living_area + tax_assessment + owner_type + stories + renovated + year_built, data =df, method="rf")
}

#   if(values()$all_comps %>% !is.null()){



sdat_models <- function(meta = NULL, comps = NULL, address = NULL){

if(is.null(meta)){
  meta = sdat_query(where=where_meta(address))
  if(nrow(meta<1)){return(NULL)}
}
comps <- filter(comps, land_use =="Residential (R)" | land_use =="Town House (TH)" , price <1500000 )
  
comps_reno <- filter(comps, renovated ==T)
mods <- list()
try({
mods$avg <- lm(price~basement, data = filter(comps, living_area > meta$living_area[[1]] -200,living_area > meta$living_area[[1]] +200) )
mods$all <- lm(price~living_area:basement:renovated, data=comps)
#mods$slr <- lm(price~living_area, data = comps_reno)
mods$lm <- lm(price~living_area:basement, data = comps_reno)
mods$lm2 <- lm(price~living_area:basement:stories, data = comps_reno)
mods$lm3 <- lm(price~living_area:basement:stories:land_use:renovated, data = comps)
mods$lm4 <- lm(price~living_area:basement:stories:neighborhood:land_use:renovated, data = comps)
#mods$rf <- train(price~living_area + basement + stories + renovated +land_use + tax_assessment, 
#                 data = comps,
#                 method="rf",
#                 ntree=50, 
#                 control = trainControl(number=1))

})
meta$renovated == T
#mods$predict <- predict(mods, meta) %>% data.frame()
return(mods)
}

sdat_predict <- function(models, meta, reno=T){
  meta$renovated <- reno
  df <- data.frame(Model = sapply(1:length(models), function(i) models[i] %>% names() %>% toupper())
                   #,Equation = sapply(1:length(models), function(i) models[[i]]$call)
                   )
  df$Estimate <- sapply(1:length(models), 
                        function(i) {
                          r <- NULL
                          try({r <- predict(models[[i]], meta)[[1]]})
                          return(r)
                        })
  
                  
  return(df)
}

library(httr)
#library(RSocrata)
library(tidyverse)
library(lubridate)
#slibrary(stringr)
library(caret)

api_endpoint = "https://opendata.maryland.gov/resource/ed4q-f8tm.json?"
api_csv = "https://opendata.maryland.gov/resource/ed4q-f8tm.csv?"

url_md = "https://opendata.maryland.gov/"
app_id = "fuixaI8OmedLW38WcI00DnoCn"

select_statement = "SELECT 
    mdp_street_address_mdp_field_address as address,
    mdp_street_address_zip_code_mdp_field_zipcode as zipcode,
    mdp_street_address_city_mdp_field_city as city,
    subdivision_code_mdp_field_subdivsn_sdat_field_37 as neighborhood,
    record_key_owner_occupancy_code_mdp_field_ooi_sdat_field_6 as owner_type,
    sales_segment_1_consideration_mdp_field_considr1_sdat_field_90 as price,
    sales_segment_2_consideration_sdat_field_110 as price2,
    sales_segment_1_transfer_date_yyyy_mm_dd_mdp_field_tradate_sdat_field_89 as date,
    sales_segment_2_transfer_date_yyyy_mm_dd_sdat_field_109 as date2,

    sales_segment_1_grantor_name_mdp_field_grntnam1_sdat_field_80 as seller,
    c_a_m_a_system_data_year_built_yyyy_mdp_field_yearblt_sdat_field_235 as year_built,
    c_a_m_a_system_data_dwelling_grade_code_and_description_mdp_field_strugrad_strudesc_sdat_field_230 as grade,
    current_assessment_year_total_assessment_sdat_field_172 as tax_assessment,
    sales_segment_3_how_conveyed_ind_sdat_field_127 as sales_type_orig,
    c_a_m_a_system_data_structure_area_sq_ft_mdp_field_sqftstrc_sdat_field_241 as living_area,
    c_a_m_a_system_data_land_area_mdp_field_landarea_sdat_field_242 as land_area,
    c_a_m_a_system_data_land_unit_of_measure_mdp_field_luom_sdat_field_243 as land_area_measure,
    land_use_code_mdp_field_lu_desclu_sdat_field_50 as land_use,
    additional_c_a_m_a_data_dwelling_construction_code_mdp_field_strucnst_sdat_field_263 as constr_code,
    additional_c_a_m_a_data_building_style_code_and_description_mdp_field_strustyl_descstyl_sdat_field_264 as bld_code_detail,
    additional_c_a_m_a_data_dwelling_type_mdp_field_strubldg_sdat_field_265 as bld_code,
    latitude_longitude as location,
    mdp_latitude_mdp_field_digycord_converted_to_wgs84 as lat,
    mdp_longitude_mdp_field_digxcord_converted_to_wgs84 as lon
"

# additional_c_a_m_a_data_building_style_code_and_description_mdp_field_strustyl_descstyl_sdat_field_1 as style_code,

select_statement = gsub("[\r\n]", "",select_statement)
select_statement = gsub("  ", "",select_statement)


where_comps = function(lat, lon, miles = 1, year = 2019){ 
  miles = miles/0.000621371 #convert to meters
  w = sprintf(" WHERE 
        
        sales_segment_1_transfer_date_yyyy_mm_dd_mdp_field_tradate_sdat_field_89 >= %s  AND 
        sales_segment_1_consideration_mdp_field_considr1_sdat_field_90 > 10000 AND 
        within_circle(latitude_longitude, %f, %f, %f)", paste0("'", year, "%'"),lat, lon, miles)
  w = gsub("[\r\n]", "",w)
  w = gsub("  ", "",w)
  return(w)
  # (sales_segment_1_consideration_mdp_field_considr1_sdat_field_90 - sales_segment_2_consideration_sdat_field_110) > 50000 AND 
  # sales_segment_2_consideration_sdat_field_110 > 0 AND 

}

clean_addresses <- function(addresses){
  props_str = toupper(addresses)
  props_str = str_replace_all(props_str,"STREET", "ST")
  props_str = str_replace_all(props_str,"AVENUE", "AVE")
  props_str = str_replace_all(props_str,"DRIVE", "DR")
  props_str = str_replace_all(props_str,"LN", "LANE")
  props_str = str_replace_all(props_str,"TERRACE", "TER")
  props_str = str_replace_all(props_str,"COURT", "CT")
  props_str = str_replace_all(props_str," RD", " ROAD")
  props_str = str_replace_all(props_str,"TRAIL", "TRL")
  props_str = str_replace_all(props_str,"PLACE", "PL")
  props_str = str_replace_all(props_str,"BOULEVARD", "BLVD")
  
  return(props_str)
}

where_meta = function(props){
  props = paste0("'", sapply(props, trimws), "'")
  props_str = paste0(props, collapse = ",")
  props_str = clean_addresses(props_str)
  
  
  
  
  w = sprintf(" WHERE mdp_street_address_mdp_field_address in (%s) ", props_str)
  w = gsub("[\r\n]", "",w)
  w = gsub("  ", "",w)
  return(w)
}

sdat_query = function(api=api_csv, select=select_statement, where,limit = 50000,return_page=0){
  query = paste0("$query= ", select, where, " LIMIT ", limit)
  full_url = paste0(api, query)
  full_url = URLencode(full_url)
  page = GET(full_url)
  if(return_page>0){return(page)}
  if(status_code(page)==200){
    frame = suppressMessages(content(page))
    
    #clean data
    try({
      frame$neighborhood <- frame$neighborhood %>% as.factor()
      frame <- filter(frame, living_area >500)
      frame$price_delta <- (frame$price - frame$price2)
      frame$date_delta <- (ymd(frame$date) - ymd(frame$date2))/ 365
      frame$renovated <- frame$price_delta > 50000 & frame$date_delta <=1 & frame$price2 >10000
      frame$basement <- grepl("with basement", tolower(frame$bld_code_detail))
      frame$stories <- sub(".*STRY *(.*?) *Story.*", "\\1", frame$bld_code_detail)
      frame$acre <- ""
      frame$acre[frame$land_area_measure =="S"] <- frame$land_area[frame$land_area_measure =="S"] / 43560
      frame$acre[frame$land_area_measure =="A"] <- frame$land_area[frame$land_area_measure =="A"]
      
    })
    return(frame)
  }else{
    return(NULL)
  }
  
}

sdat_q=function(api=api_csv, q){
  query= paste0("$q=", q)
  full_url = paste0(api, query)
  full_url = URLencode(full_url)
  page = GET(full_url)
  if(status_code(page)==200){
    frame = content(page)
    return(frame)
  }else{
    return(NULL)
  }
  
}
sdat_get_meta <- function(df, col = "address"){
  results <- NULL
  for(i in 1:nrow(df)){
    tmp_result <- NULL
    tmp_result <- sdat_query(where=where_meta(df[,col][i]))
    if(!is.null(tmp_result)){
      #tmp_result$search_address <- df[,col][i] 
      if(is.null(results)){
        results <- tmp_result
      }else{
        results <- rbind(results, tmp_result)
    }
  }
  }
  df$search_address <- clean_addresses(df[,col])
  df <- left_join(df, results, by=c("search_address" = "address"))
  return(df)
}

sdat_get_many_properties = function(addresses){
  n = length(addresses)
  result = NULL
  beg = 0
  end = 0
  while(end<n){
    beg= beg+1
    end = min(beg +9 , n)
    print(paste("beginning", beg))
    print(paste("END", end))
    tmp_result = sdat_query(where=where_meta(addresses[beg:end]))
    if(is.null(result)){
      result = tmp_result
    }else{
      result = rbind(result, tmp_result)
    }
    tmp_result = NULL   
    beg = end
  }
  return(result)
}

model.lm <- function(df, reno = T, dist = 1){
  if(reno){df <- filter(df, renovated==T)}
  
  lm(price~living_area + stories + basement , data = df)
}

model.rf <- function(df){
  #control <- trainControl(method='boot', 
  #                        number=10, 
  #                       repeats=3)
  train(price~living_area + tax_assessment + owner_type + stories + renovated + year_built, data =df, method="rf")
}

#   if(values()$all_comps %>% !is.null()){



sdat_models <- function(meta = NULL, comps = NULL, address = NULL){

if(is.null(meta)){
  meta = sdat_query(where=where_meta(address))
  if(nrow(meta<1)){return(NULL)}
}
comps <- filter(comps, land_use =="Residential (R)" | land_use =="Town House (TH)" , price <1500000 )
  
comps_reno <- filter(comps, renovated ==T)
mods <- list()
try({
mods$avg <- lm(price~basement, data = filter(comps, living_area > meta$living_area[[1]] -200,living_area > meta$living_area[[1]] +200) )
mods$all <- lm(price~living_area:basement:renovated, data=comps)
#mods$slr <- lm(price~living_area, data = comps_reno)
mods$lm <- lm(price~living_area:basement, data = comps_reno)
mods$lm2 <- lm(price~living_area:basement:stories, data = comps_reno)
mods$lm3 <- lm(price~living_area:basement:stories:land_use:renovated, data = comps)
mods$lm4 <- lm(price~living_area:basement:stories:neighborhood:land_use:renovated, data = comps)
#mods$rf <- train(price~living_area + basement + stories + renovated +land_use + tax_assessment, 
#                 data = comps,
#                 method="rf",
#                 ntree=50, 
#                 control = trainControl(number=1))

})
meta$renovated == T
#mods$predict <- predict(mods, meta) %>% data.frame()
return(mods)
}

sdat_predict <- function(models, meta, reno=T){
  meta$renovated <- reno
  df <- data.frame(Model = sapply(1:length(models), function(i) models[i] %>% names() %>% toupper())
                   #,Equation = sapply(1:length(models), function(i) models[[i]]$call)
                   )
  df$Estimate <- sapply(1:length(models), 
                        function(i) {
                          r <- NULL
                          try({r <- predict(models[[i]], meta)[[1]]})
                          return(r)
                        })
  
                  
  return(df)
}

File: sdat.py
#Get data
from sodapy import Socrata
import pandas as pd
import difflib
pd.options.display.max_columns = 500


url_md = "opendata.maryland.gov"
#id_md = "pwpp-tk9r"
id_md ="ed4q-f8tm"
app_id = "fuixaI8OmedLW38WcI00DnoCn"

select_statement = '''
    mdp_street_address_mdp_field_address as address,
    mdp_street_address_zip_code_mdp_field_zipcode as zipcode,
    mdp_street_address_city_mdp_field_city as city,
    subdivision_code_mdp_field_subdivsn_sdat_field_37 as neighborhood,
    record_key_owner_occupancy_code_mdp_field_ooi_sdat_field_6 as owner_type,
    sales_segment_1_consideration_mdp_field_considr1_sdat_field_90 as price,
    sales_segment_2_consideration_sdat_field_110 as price2,
    sales_segment_1_transfer_date_yyyy_mm_dd_mdp_field_tradate_sdat_field_89 as date,
    sales_segment_2_transfer_date_yyyy_mm_dd_sdat_field_109 as date2,
    sales_segment_1_grantor_name_mdp_field_grntnam1_sdat_field_80 as seller,
    c_a_m_a_system_data_year_built_yyyy_mdp_field_yearblt_sdat_field_235 as year_built,
    c_a_m_a_system_data_dwelling_grade_code_and_description_mdp_field_strugrad_strudesc_sdat_field_230 as grade,
    current_assessment_year_total_assessment_sdat_field_172 as tax_assessment,
    sales_segment_3_how_conveyed_ind_sdat_field_127 as sales_type_orig,
    c_a_m_a_system_data_structure_area_sq_ft_mdp_field_sqftstrc_sdat_field_241 as living_area,
    c_a_m_a_system_data_land_area_mdp_field_landarea_sdat_field_242 as land_area,
    c_a_m_a_system_data_land_unit_of_measure_mdp_field_luom_sdat_field_243 as land_area_measure,
    land_use_code_mdp_field_lu_desclu_sdat_field_50 as land_use,
    additional_c_a_m_a_data_dwelling_construction_code_mdp_field_strucnst_sdat_field_263 as constr_code,
    additional_c_a_m_a_data_building_style_code_and_description_mdp_field_strustyl_descstyl_sdat_field_264 as bld_code_detail,
    additional_c_a_m_a_data_dwelling_type_mdp_field_strubldg_sdat_field_265 as bld_code,
    latitude_longitude as location,
    mdp_latitude_mdp_field_digycord_converted_to_wgs84 as lat,
    mdp_longitude_mdp_field_digxcord_converted_to_wgs84 as lon
'''
def clean_addresses(addresses):
    props_str = addresses.upper()
    props_str = props_str.replace("STREET", "ST")
    props_str = props_str.replace("AVENUE", "AVE")
    props_str = props_str.replace("DRIVE", "DR")
    props_str = props_str.replace("LN", "LANE")
    props_str = props_str.replace("TERRACE", "TER")
    props_str = props_str.replace("COURT", "CT")
    props_str = props_str.replace(" RD", " ROAD")
    props_str = props_str.replace("TRAIL", "TRL")
    props_str = props_str.replace("PLACE", "PL")
    props_str = props_str.replace("BOULEVARD", "BLVD")
    
    return props_str



def property_metadata(address, save_csv = False, select = select_statement):
    '''
        Query SDAT
        Return dictionary with metadata
    '''
    client = Socrata(url_md, app_id)
    address = clean_addresses(address)
    results = client.get(id_md, 
                     select =select,
                     where ="mdp_street_address_mdp_field_address='{}'".format(address))
    
    df = pd.DataFrame(results)
    closest = difflib.get_close_matches(address.upper(), df.address, n=1)
    #print('Closest Property:', closest[0])
    df = df[df.address == closest[0]]
    #df['lon'] = df.location.apply(lambda x: x['coordinates'][0])
    #df['lat'] = df.location.apply(lambda x: x['coordinates'][1])
    client.close()
    if save_csv:
            sdat = pd.read_csv('sdat_properties.csv')
            #sdat = pd.concat([sdat, df], ignore_index=True)
            result = pd.concat([sdat, df])
            result.to_csv('sdat_properties.csv', index=False)
    return df            

def miles_to_meter(miles):
    return float(miles /0.000621371)

def get_params(df, miles = 1):
    d = dict()
    d['miles'] = miles_to_meter(miles)
    for col in df.columns:
        d[col] = df[col].iloc[0]
    return d

def get_query(df, miles =1):
    q_params = get_params(df, miles)
    query = 'SELECT ' + select_statement +  '''
    
    WHERE  
        sales_segment_1_consideration_mdp_field_considr1_sdat_field_90 > 0 AND
        sales_segment_2_consideration_sdat_field_110 > 0 AND
        within_circle(latitude_longitude, {lat}, {lon}, {miles})


    LIMIT 10000
    '''.format(**q_params)
    return query

def get_comps_sdat(address, miles=1, save_csv=False):
    meta_df = property_metadata(address)
    query = get_query(meta_df, miles)
    client = Socrata(url_md, app_id)
    response = client.get(id_md, query = query)
    df = pd.DataFrame(response)
    num = []
    non_num = []
    for col in df.columns:
        try:
            df[col] = pd.to_numeric(df[col])
            num.append(col)
        except:
            non_num.append(col)

    df['id'] = meta_df.address.iloc[0]
    df.date = pd.to_datetime( df['date'])
    df.date2= pd.to_datetime(df['date2'])
    df['basement'] = df['style_code'].apply(lambda x: 'with basement' in x.lower())
    df['price_change'] = df['price'].astype(float) - df['price2'].astype(float)
    df['date_change'] = (df.date - df.date2).dt.days
    df['norm_price'] = ((1+(0.03/365)) ** (pd.datetime.today() - df['date']).dt.days) * df.price
    client.close()
    if save_csv: df.to_csv(meta_df.address.iloc[0] +".csv", index = False)
    results = (meta_df, df)
    return results #df

def filter_comps(df, price = 75000, turn_around = 400, t_time = pd.to_datetime('2018-01-01')):
    df_copy = df[
    (df['price_change']>=price) & 
    (df['date_change']<=turn_around) & 
    (df.date > pd.to_datetime(t_time))
    ]
    return df_copy


def find_comps(prop, update=True):
    if update:
        done_comps = []
    else:
        comps_df =  pd.read_csv('comps.csv')
        done_comps = comps_df.id.unique()        
    try:
        meta, comp = get_comps_sdat(prop)
        if comp.id.iloc[0] in done_comps:
            #do nothing
            print(prop, 'already done')
            filt_comp = filter_comps(comps_df.query('id == {}'.format(comps.id.iloc[0])))
            
        else:
            #load comp to csv
            filt_comp = filter_comps(comp)
            #filt_comp.to_csv('comps.csv', mode="a", header=False, index = False)

    except:
        search_address = prop.split()[:-1]
        search_address = ' '.join(search_address)
        try:
            if comp.id.iloc[0] in done_comps:
                #do nothing
                print(prop, 'already done')
            else:
                meta, comp = get_comps_sdat(search_address)
                filt_comp = filter_comps(comp, price=50000)
                #filt_comp.to_csv('comps.csv', mode="a", header=False, index = False)
        except:
            print('{}: Could not find property'.format(prop))
            return None
    
    results = (meta, filt_comp)    
    return results #filt_comp

def potential_property(f=None, df=None, address_col='address'):
    '''
    Function will create comps for each property in csv
    
    PARAMS:
        f: filename
        df: optional parameter to provide dataframe
        address_col: name of column that contains address
    '''
    if isinstance(df, pd.DataFrame): pass
    else: df = pd.read_csv(f)
    df.columns = df.columns.str.lower()
    address_col = address_col.lower()
    comps = pd.read_csv('comps.csv')
    done_comps = [] #comps.id.unique()
    
    properties = pd.read_csv('properties.csv')
    
    
    
    for prop in df[address_col].unique():
        comps = find_comps(prop)
        if isinstance(comps, pd.DataFrame):
            #do something
            vals = {'Address': [comps.id.iloc[0]],
                    'Re-Sale Price': [comps['price'].median()], 
                    'List Price': [df[df['address'] == prop]['current price'].mean()] 
                   }
            tmp_prop = pd.DataFrame(vals)
            tmp_prop['Margin'] = tmp_prop['Re-Sale Price'] - tmp_prop['List Price']
            properties = properties.append(tmp_prop, ignore_index=True)
            tmp_prop = None
        else:
            print(prop, 'could not be added')
        comps = None
    return properties
